Rewrite JavaScriptCompiler to use modern JVM features (Java 17)

uschindler opened the pull request:
This PR rewrites the JavascriptCompiler for the main branch to use more modern JVM features to be easier configurable and does not need a separate classloader per expression (uses hidden classes).

Hidden classes are a new feature of Java 15: It allows to define a hidden class with byte code without using a classloader using the access constraints of a given private lookup. This is the same technique used by lambdas and other invokeDynamic features and replaces the old Unsafe API that Lucene never used. The downsides with a separate on-class classloader are described in #8933. Mainly with many expressions compiled in parallel there's a lock in parent classloader while looking up methods/fields and loading classes.

The new Java 15 hidden classes are unloaded by the JVM if the get out of scope, so the workaround with a separate classloader is no longer needed.

A second reason why separate classloaders were needed is to allow code access functions form other classloaders. This stopped me doing the whole optimization at first until I rewrote the whole function lookup to be more modern:

- The function array is now a `Map<String, MethodHandle>`. This allows to create arbitrary functions anyhere in code like Elasticsearch/Opensearch (the code may even be private or in another classloader). The access to those method is no longer the deal of the compiler, if the MethodHandle is there, it can be used. Access is the deal of the user of the expressions API. In addition you can also use "synthetic" methodhandles like those generated by `MethodHandles#filterArguments` or you can use non-static methods as long as you bind the receiver to a fixed instance. Some examples were added to the tests.
- The problem with MethodHandles is that they should be `static final` constants in the class or must appear in the constant pool of the class. This is now the second trick used here: Since Java 11 there is the feature of dynamic constants: Basically the LDC instruction can load a MethodHandle constant (which is treated as the "perfect" MethodHandle for Hotspot) using a bootstrap method (invokedynamic v2). The Javascript compiler now implements a bootstrap method that looks up the function dynamically from the above `Map<String,MethodHandle>`.
- Java 16 brought another feature that allows to add `classData` to hidden classes (Java 15). This classData is saved in the private bootstrapper `Lookup` and is accessible during bootstrapping constants. When loading the hidden class we pass our function map as classdata to the lookup and this allows the constant bootstrapping to return the MethodHandle.
- I have to do more benchmarking, but generally the dynamic constants and using LDC and `MethodHandle.invokeExact()` should be enough to call ALL functions. For now I added a shortcut: If the `MethodHandle` in our map links directly to a static Method in a public class in same classloader, the bytecode generator for now uses the `MethodHandleInfo` to generate a conventional `invokestatic`. I plan to remove this "optimization" after more benchmarking. You can easily disable the optimization by setting `directFunctions` to an empty Map.

There is one downside in the hidden class feature that I added a workaround for: Hidden classes never appear in stack traces so my initial version did not print the actual JS function in stack trace. I worked around this by adding "exception patching" in the base class. It catches all Throwables throws by the generated code in the subclass and changes the stack trace to include the source code and hidden class name. This allows same user experience before.

The current code is not backward compatible and needs an entry in `MIGRATE.txt`. I will investiagte how to add some helper methods for those people using `java.lang.reflect.Method` maps. Those can be easily converted to `MethodHandle` with some wrapper method. Support for separate classloader was removed. This is not a security issue, because the methods available for calling are defined externally and the API user is still responsible to only supply methods which should be accessible by users. The access checking is done by the provider of function map. For safety you can still let run whole of Lucene or the expressions module in a separate module/classloader.

@rjernst, @jdconrad: Maybe this is also an idea for Painless.

This PR replaces the outdated https://github.com/apache/lucene-solr/pull/1837. This closes #8933.

uschindler committed:
Rewrite Javascript expression compiler to use hidden classes and MethodHandles for functions

uschindler committed:
Use dynamic constants for MethodHandles

uschindler commented on lucene/expressions/src/java/org/apache/lucene/expressions/js/JavascriptCompiler.java:
```diff
@@ -201,48 +210,75 @@ private static void unusedTestCompile() throws IOException {
+        map.put(e.getKey(), info);
+      }
+    }
+    this.directFunctions = Collections.unmodifiableMap(map);
```
set this to `Map.empty()` to test using dynamic constants and `MethodHandles#invokeExact` for all functions.

uschindler committed:
Remove invokestatic code and handle everything through dynamic constants

uschindler committed:
Rewrite code to patch stack trace (keep Expressions class unmodified)

uschindler commented:
The latest commit changed the stack trace patching to be implemented inside the compiled expression. The original benchmarks had shown that the extra `evaluate()` method in the base class caused some slowdown. By doing the stack trace patching the "new way", all is sane again. Due to the change the original `Expression` class is now unmodified.

rmuir commented:
I like the change!

Can be a followup ticket, but i think we should remove `ClassLoader` usage from `TestCustomFunctions.java`. It doesn't need to use classloader to test anymore, since it can just pass a MethodHandle: so maybe the test is easy to fix. With that test fixed, `createClassLoader` permission can be removed from tests, lucene code really shouldn't be messing with classloaders if we have a more optimal way?

uschindler committed:
Improve generating of constant names

uschindler commented:
I will also post my benchmark, as it allows to test the performance of expressions.

uschindler committed:
Remove classloader test (no longer needed)

uschindler committed:
Add benchmark

uschindler commented:
Benchmark was added and classloader code removed.

I will add a helper function to support some compatibility with legacy code still using `java.lang.reflect.Method`, but after that it's ready. I will also add a MIGRATE information how to convert old code.

uschindler committed:
use better exception in benchmark

uschindler commented:
While writing the benchmark I figured out that the Expression base class has some problems with a missing `IOException` on the `evaluate(DoubleValues[])` method. As bytecode knows no exceptions in throws clauses, the evaluation method may throw `IOException` unexpected because `DoubleValues#doubleValue()` is defined to throw `IOException`. This is a bug and should be fixed separately. I will open an issue to fix this in main.

uschindler committed:
Add documentation, migration guide and a utility method to convert legacy function maps

uschindler commented:
I added documentation and migration guide.

uschindler committed:
also ignore SecurityException here while checking compatibility (if it happens only an imprecise error message is thrown)

uschindler committed:
Use Map.copyOf to not clone the map each time we compile an expression

uschindler committed:
Add another test with same method multiple times

uschindler committed:
Update ASM to 9.6 and set classfile version to Java 17

uschindler committed:
Cleanup classloader permissions, unfortunately "createClassLoader" is still needed for Jacoco for God knows what

rmuir commented on lucene/expressions/src/java/org/apache/lucene/expressions/js/JavascriptCompiler.java:
```diff
@@ -201,48 +226,57 @@ private static void unusedTestCompile() throws IOException {
+      throws ReflectiveOperationException {
+    final MethodHandle ctor = lookup.findConstructor(expressionClass, MT_EXPRESSION_CTOR_LOOKUP);
+    try {
+      return (Expression) ctor.invoke(sourceText, externalsMap.keySet().toArray(String[]::new));
```
any easy way to change this to `invokeExact()` ? I realize its just for creating the expression but you know how these things go. I tried changing it locally and got WrongMethodTypeException so some adaptation is happening....

uschindler replied:
this is a one-time call and manually adapting makes no sense here. We have more of those in Lucene.

You can adapt it, but this is not worth it.

uschindler commented:
> I like the change!
> 
> Can be a followup ticket, but i think we should remove `ClassLoader` usage from `TestCustomFunctions.java`. It doesn't need to use classloader to test anymore, since it can just pass a MethodHandle: so maybe the test is easy to fix. With that test fixed, `createClassLoader` permission can be removed from tests, lucene code really shouldn't be messing with classloaders if we have a more optimal way?

I cleaned up the policy files, but the `createClassLoader` permission is still needed by the the Jacoco agent, which uses no AccessController.... So I can't limit access. But I moved the permission to the Jacoco ones. We may open a bug report for Jacoco to fix this.

I was successful with removing `getClassLoader` permission. This one is no longer needed as we are no longer inspecting classloaders of unknown classes. I think this permission can also be removed in 9.x.

Finally I removed some permission from replication in main branch caused by servlet API.

rmuir commented:
some of the permissions involved are probably only safe to remove in 10.x due to #12038

rmuir reviewed (approved):
Thanks for modernizing this, no more classloaders or reflection :)

rjernst commented:
Thanks for the ping @uschindler , great idea! I opened https://github.com/elastic/elasticsearch/issues/102955 to track the same work in Elasticsearch.
